zerver/lib/streams.py

```diff
@@ -642,3 +643,18 @@ def get_stream_by_narrow_operand_access_unchecked(operand: Union[str, int], real
     if isinstance(operand, str):
         return get_stream(operand, realm)
     return get_stream_by_id_in_realm(operand, realm)
+
+
+def get_subscribed_streams_for_user_by_history_public_to_subscribers(
+    realm: Realm, user_profile: Optional[UserProfile], history_public_to_subscribers: bool
+) -> "QuerySet[Stream]":
+
+    assert user_profile is not None
+
+    subscribed_stream_ids = get_subscribed_stream_ids_for_user(user_profile)
+
+    return Stream.objects.filter(
+        realm=realm,
+        id__in=subscribed_stream_ids,
+        history_public_to_subscribers=history_public_to_subscribers,
+    )
```

Comment:
This function looks correct to me, but we probably want some tests for it.  One possibility is that we feel good about the end-to-end `message_fetch` tests covering this well; but another good option would be to modify existing `test_subs.py` tests (search for `history_public_to_subscribers` ) to call this function and make a few assertions about the result.  Could be worth doing in this commit.